File: src/LeetCode/Algorithms/Java/Binary_Tree_Preorder_Traversal.java
package LeetCode.Algorithms.Java;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

// https://leetcode.com/problems/binary-tree-preorder-traversal/
public class Binary_Tree_Preorder_Traversal {
	public static class TreeNode {
		int val;
		TreeNode left;
		TreeNode right;
		
		TreeNode(int x) {
			this.val = x;
		}
	}
	
	public static void main(String[] args){
		TreeNode root = new TreeNode(1);
		root.left = null;
		root.right = new TreeNode(2);
		root.right.left = new TreeNode(3);
		
		List<Integer> list = preorderTraversal(root);
		
		for (int i : list) {
			System.out.print(i + " ");
		}
		System.out.println();
	}
	
	public static List<Integer> preorderTraversal(TreeNode root) {
		ArrayList<Integer> list = new ArrayList<Integer>();
		Stack<TreeNode> s = new Stack<TreeNode>();
		
		if(root != null) {
			s.push(root);
			
			while(!s.isEmpty()) {
				TreeNode tn = s.pop();
				list.add(tn.val);
				
				if(tn.right != null) {
					s.push(tn.right);
				}
				
				if(tn.left != null) {
					s.push(tn.left);
				}
			}
		}
		
		return list;
	}
}

File: src/LeetCode/Algorithms/JavaScript/Binary_Tree_Preorder_Traversal.js
/**
 * https://leetcode.com/problems/binary-tree-preorder-traversal/
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[]}
 */
var preorderTraversal = function(root) {
	let result = [];
	
	if (root != null) {
		result.push(root.val);
		result.push(...preorderTraversal(root.left));
		result.push(...preorderTraversal(root.right));
	}
	
	return result;
	
	/*let result = [];
	let stack = [root];
	
	if (root != null) {
		while (stack.length) {
			const curr = stack.pop();
			
			if (curr) {
				result.push(curr.val);
				
				stack.push(curr.right);
				stack.push(curr.left);
			}
		}
	}
	
	return result;*/
};

function TreeNode(val) {
	this.val = val;
	this.left = this.right = null;
}

let root = new TreeNode(1);

root.right = new TreeNode(2);
root.right.left = new TreeNode(3);

console.log(preorderTraversal(root));